From 620a5e0831b8da68dd83b98fa96540b7ba3849a4 Mon Sep 17 00:00:00 2001
From: "Guoniu.Zhou" <guoniu.zhou@nxp.com>
Date: Thu, 31 May 2018 15:10:14 +0800
Subject: [PATCH 3903/5242] MLK-18440: ISI: fix the first frame black issue

commit  4eac33377456ec7b30c8948bf954107574ed42dc from
https://source.codeaurora.org/external/imx/linux-imx.git

Camera sensor need to be enabled before ISI when
resolution of caputure image is less than 640*480,
but hdmi need to be enabled after ISI, so add this
patch to compatible these two cases

Signed-off-by: Guoniu.Zhou <guoniu.zhou@nxp.com>
(cherry picked from commit c8d17ffcb0ab007d0be228b2cc4562b9cbc07501)
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/media/platform/imx8/mxc-isi-cap.c |   26 ++++++++++++++++++--------
 1 file changed, 18 insertions(+), 8 deletions(-)

diff --git a/drivers/media/platform/imx8/mxc-isi-cap.c b/drivers/media/platform/imx8/mxc-isi-cap.c
index e0c7cde..4603977 100644
--- a/drivers/media/platform/imx8/mxc-isi-cap.c
+++ b/drivers/media/platform/imx8/mxc-isi-cap.c
@@ -881,11 +881,16 @@ static int mxc_isi_cap_streamon(struct file *file, void *priv,
 	int ret;
 
 	dev_dbg(&mxc_isi->pdev->dev, "%s\n", __func__);
-	mxc_isi_channel_enable(mxc_isi);
 
-	ret = vb2_ioctl_streamon(file, priv, type);
-
-	mxc_isi_pipeline_enable(mxc_isi, 1);
+	if (mxc_isi->interface[IN_PORT] == ISI_INPUT_INTERFACE_HDMI) {
+		mxc_isi_channel_enable(mxc_isi);
+		ret = vb2_ioctl_streamon(file, priv, type);
+		mxc_isi_pipeline_enable(mxc_isi, 1);
+	} else {
+		mxc_isi_pipeline_enable(mxc_isi, 1);
+		ret = vb2_ioctl_streamon(file, priv, type);
+		mxc_isi_channel_enable(mxc_isi);
+	}
 
 	return ret;
 }
@@ -897,11 +902,16 @@ static int mxc_isi_cap_streamoff(struct file *file, void *priv,
 	int ret;
 
 	dev_dbg(&mxc_isi->pdev->dev, "%s\n", __func__);
-	mxc_isi_channel_disable(mxc_isi);
 
-	ret = vb2_ioctl_streamoff(file, priv, type);
-
-	mxc_isi_pipeline_enable(mxc_isi, 0);
+	if (mxc_isi->interface[IN_PORT] == ISI_INPUT_INTERFACE_HDMI) {
+		mxc_isi_channel_disable(mxc_isi);
+		ret = vb2_ioctl_streamoff(file, priv, type);
+		mxc_isi_pipeline_enable(mxc_isi, 0);
+	} else {
+		mxc_isi_pipeline_enable(mxc_isi, 0);
+		ret = vb2_ioctl_streamoff(file, priv, type);
+		mxc_isi_channel_disable(mxc_isi);
+	}
 
 	return ret;
 }
-- 
1.7.9.5

